lib/rmt.toit

```diff
@@ -0,0 +1,181 @@
+    this.period = period & 0x7FFF
+    this.level = level & 0b1
+
+  /** Deserialized an item from the given $bytes at the given $index. */
+  constructor.from_bytes index/int bytes/ByteArray:
+    period = bytes[index] | ((bytes[index + 1] & 0x7F) << 8)
+    level = bytes[index + 1] >> 7
+
+  first_byte_ -> int:
+    return period & 0xFF
+
+  second_byte_ -> int:
+    return (period >> 8 ) | (level << 7)
+
+  /** See $super. */
+  operator == other/any:
+    if other is not Item: return false
+
+    return level == other.level and period == other.period
+
+  /** See $super. */
+  stringify -> string:
+    return "($period, $level)"
+
+/**
+An RMT channel.
+
+The channel must be configured after construction.
+
+  The channel can be configured for either RX or TX.
+*/
+class Channel:
+  num/int
+  pin/gpio.Pin
+
+  res_/ByteArray? := null
+
+  /** Constructs a channel using the given $num using the given $pin. */
+  constructor .pin .num:
+    res_ = rmt_use_ resource_group_ num
+
+  /**
+  Configure the channel for RX.
+
+  - $mem_block_num is the number of memory blocks (512 bytes) used by this channel.
+  - $clk_div is the source clock divider. Must be in the range [0,255].
+  - $flags is the configuration flags. See the ESP-IDF documentation for available flags.
+  - $idle_threshold is the amount of clock cycles the receiver will run without seeing an edge.
+  - $filter_en whether the filter is enabled.
+  - $filter_ticks_thresh pulses shorter than this value is filtered away.
+    Only works with $filter_en. The value must be in the range [0,255].
+  */
+  config_rx
+      --mem_block_num/int=1
+      --clk_div/int=80
+      --flags/int=0
+      --idle_threshold/int=12000
+      --filter_en/bool=true
+      --filter_ticks_thresh/int=100:
+    rmt_config_rx_ pin.num num mem_block_num clk_div flags idle_threshold filter_en filter_ticks_thresh
+
+  /**
+  Configure the channel for TX.
+
+  - $mem_block_num is the number of memory blocks (512 bytes) used by this channel.
+  - $clk_div is the source clock divider. Must be in the range [0,255].
+  - $flags is the configuration flags. See the ESP-IDF documentation for available flags.
+  - $carrier_en
+  */
+  config_tx
+      --mem_block_num/int=1
+      --clk_div/int=80
+      --flags/int=0
+      --carrier_en/bool=false
+      --carrier_freq_hz/int=38000
+      --carrier_level/int=1
+      --carrier_duty_percent/int=33
+      --loop_en/bool=false
+      --idle_output_en/bool=true
+      --idle_level/int=0:
+    rmt_config_tx_ pin.num num mem_block_num clk_div flags carrier_en carrier_freq_hz carrier_level carrier_duty_percent loop_en idle_output_en idle_level
+
+  close:
+    if res_:
+      rmt_unuse_ resource_group_ res_
+      res_ = null
+
+class Controller:
+  static transfer channel/Channel items/List/*<Item>*/:
+    rmt_transfer_ channel.num
+      items_to_bytes_ items
+
+  static transfer_and_read --rx/Channel --tx/Channel items/List/*<Item>*/ max_items_size/int -> List:
```

Comment:
You should be able to use the same pin, but it should have two different channels.

I have thought about the same thing wrt items. For something like 1-wire protocol it would take one item per bit you want to send. That adds up very quickly! I think a wrapper class of easy access would still be a good idea.